src/MichiWalletReceiptNFT.sol

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+contract MichiWalletReceiptNFT is ERC721, AccessControl {
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+
+    /// @notice tracks the total supply minted
+    uint256 public totalSupply;
+
+    error UnauthorizedMinter(address user);
+
+    event Mint(address indexed minter, uint256 nftId);
+
+    constructor() ERC721("Michi Wallet Receipt NFT", "MICHI RECEIPT") {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function mint(address to, uint256 tokenId) external onlyRole(MINTER_ROLE) {
+        _safeMint(to, tokenId);
+
+        emit Mint(to, tokenId);
+        totalSupply++;
```

Comment:
its best to get in the habit of checks effects interactions. increase total supply before mint

Reply:
fixed